server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -112,14 +113,15 @@ private void setMaxConcurrentFetches(Integer maxConcurrentFetches) {
     @Override
     public SnapshotShardSizeInfo snapshotShardSizes() {
         synchronized (mutex) {
-            final ImmutableOpenMap.Builder<SnapshotShard, Long> snapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShards);
+            Map<SnapshotShard, Long> snapshotShardSizes = knownSnapshotShards;
             if (failedSnapshotShards.isEmpty() == false) {
+                snapshotShardSizes = new HashMap<>(snapshotShardSizes); // create a new map to modify
                 for (SnapshotShard snapshotShard : failedSnapshotShards) {
                     Long previous = snapshotShardSizes.put(snapshotShard, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
                     assert previous == null : "snapshot shard size already known for " + snapshotShard;
                 }
             }
-            return new SnapshotShardSizeInfo(snapshotShardSizes.build());
+            return new SnapshotShardSizeInfo(snapshotShardSizes);
```

Comment:
Should we be passing in here Collections.unmodifiableMap(snapshotShardSizes) instead of the mutable HashMap?

Reply:
Probably - but I'm not sure if the immutable-ness should be handled by the `SnapshotShardSizeInfo` constructor, as it's a public method/class (cf `SnapshotsInProgress.Entry`). How public is this class?